evaluator/builtin_time.go

```diff
@@ -673,6 +675,328 @@ func builtinFromUnixTime(args []types.Datum, _ context.Context) (d types.Datum,
+	if !succ {
+		return false
+	}
+
+	return strToDate(t, dateRemain, formatRemain)
+}
+
+// getFormatToken takes one format control token from the string.
+// format "%d %H %m" will get token "%d" and the remain is " %H %m".
+func getFormatToken(format string) (token string, remain string, succ bool) {
+	if len(format) == 0 {
+		return "", "", true
+	}
+
+	// Just one character.
+	if len(format) == 1 {
+		if format[0] == '%' {
+			return "", "", false
+		}
+		return format, "", true
+	}
+
+	// More than one character.
+	if format[0] == '%' {
+		return format[:2], format[2:], true
+	}
+
+	return format[:1], format[1:], true
+}
+
+func skipWhiteSpace(input string) string {
+	for i, c := range input {
+		if !unicode.IsSpace(c) {
+			return input[i:]
+		}
+	}
+	return ""
+}
+
+var weekdayAbbrev = map[string]time.Weekday{
+	"Sun": time.Sunday,
+	"Mon": time.Monday,
+	"Tue": time.Tuesday,
+	"Wed": time.Wednesday,
+	"Thu": time.Tuesday,
+	"Fri": time.Friday,
+	"Sat": time.Saturday,
+}
+
+var monthAbbrev = map[string]time.Month{
+	"Jan": time.January,
+	"Feb": time.February,
+	"Mar": time.March,
+	"Apr": time.April,
+	"May": time.May,
+	"Jun": time.June,
+	"Jul": time.July,
+	"Aug": time.August,
+	"Sep": time.September,
+	"Oct": time.October,
+	"Nov": time.November,
+	"Dec": time.December,
+}
+
+type dateFormatParser func(t *time.Time, date string) (remain string, succ bool)
+
+var dateFormatParserTable = map[string]dateFormatParser{
+	"%a": abbreviatedWeekday,
+	"%b": abbreviatedMonth,
+	"%c": monthNumeric,
+	"%D": dayOfMonthWithSuffix,
+	"%Y": yearNumericFourDigits,
+	"%m": monthNumericTwoDigits,
+	"%d": dayOfMonthNumericTwoDigits,
+	"%H": hour24TwoDigits,
+	"%i": minutesNumeric,
+	"%s": secondsNumeric,
+}
+
+func matchDateWithToken(t *time.Time, date string, token string) (remain string, succ bool) {
+	if parse, ok := dateFormatParserTable[token]; ok {
+		return parse(t, date)
+	}
+
+	if strings.HasPrefix(date, token) {
+		return date[len(token):], true
+	}
+	return date, false
+}
+
+func parseTwoDigits(input string) (int, bool) {
+	if len(input) < 2 {
+		return 0, false
+	}
+
+	v, err := strconv.ParseUint(input[:2], 10, 64)
+	if err != nil {
+		return int(v), false
+	}
+	return int(v), true
+}
+
+func hour24TwoDigits(t *time.Time, input string) (string, bool) {
+	v, succ := parseTwoDigits(input)
+	if !succ || v >= 24 {
+		return input, false
+	}
+	timeSetHour(t, v)
+	return input[2:], true
+}
+
+func secondsNumeric(t *time.Time, input string) (string, bool) {
+	v, succ := parseTwoDigits(input)
+	if !succ || v >= 60 {
+		return input, false
+	}
+	timeSetSecond(t, v)
+	return input[2:], true
+}
+
+func minutesNumeric(t *time.Time, input string) (string, bool) {
+	v, succ := parseTwoDigits(input)
+	if !succ || v >= 60 {
+		return input, false
+	}
+	timeSetMinute(t, v)
+	return input[2:], true
+}
+
+func dayOfMonthNumericTwoDigits(t *time.Time, input string) (string, bool) {
+	v, succ := parseTwoDigits(input)
```

Comment:
mysql behavior:

```
mysql> select str_to_date('2016 02 31', '%Y %m %D');
+---------------------------------------+
| str_to_date('2016 02 31', '%Y %m %D') |
+---------------------------------------+
| 2016-02-31                            |
+---------------------------------------+
1 row in set (0.00 sec)
```

mysql permit invalid date, while Go can't...and that why we need refact.
Of course, I'll be glad if you also review this PR https://github.com/pingcap/tidb/pull/2098